--- libunwind-1.1/src/Makefile.am.orig	2012-10-06 06:54:38.000000000 +0200
+++ libunwind-1.1/src/Makefile.am	2012-10-12 15:56:24.425745356 +0200
@@ -39,6 +39,7 @@
 	ptrace/_UPT_find_proc_info.c ptrace/_UPT_get_dyn_info_list_addr.c \
 	ptrace/_UPT_put_unwind_info.c ptrace/_UPT_get_proc_name.c	  \
 	ptrace/_UPT_reg_offset.c ptrace/_UPT_resume.c
+libunwind_ptrace_la_LIBADD = libunwind-$(arch).la
 noinst_HEADERS += ptrace/_UPT_internal.h
 
 ### libunwind-coredump:
@@ -58,6 +59,7 @@
 	coredump/_UPT_resume.c
 libunwind_coredump_la_LDFLAGS = $(COMMON_SO_LDFLAGS) \
 				-version-info $(COREDUMP_SO_VERSION)
+libunwind_coredump_la_LIBADD = libunwind-$(arch).la
 noinst_HEADERS += coredump/_UCD_internal.h
 
 ### libunwind-setjmp:
