--- libunwind-0.98.5/include/dwarf_i.h.orig	2006-02-26 12:55:00.000000000 +0100
+++ libunwind-0.98.5/include/dwarf_i.h	2006-02-26 12:56:43.888585250 +0100
@@ -29,17 +29,17 @@
 typedef union
   {
     int8_t s8;
-    int16_t s16;
-    int32_t s32;
-    int64_t s64;
+    int16_t s16		__attribute__ ((packed));
+    int32_t s32		__attribute__ ((packed));
+    int64_t s64		__attribute__ ((packed));
     uint8_t u8;
-    uint16_t u16;
-    uint32_t u32;
-    uint64_t u64;
-    unw_word_t w;
+    uint16_t u16	__attribute__ ((packed));
+    uint32_t u32	__attribute__ ((packed));
+    uint64_t u64	__attribute__ ((packed));
+    unw_word_t w	__attribute__ ((packed));
     void *ptr;
   }
-dwarf_misaligned_value_t __attribute__ ((packed));
+dwarf_misaligned_value_t;
 
 static inline int
 dwarf_reads8 (unw_addr_space_t as, unw_accessors_t *a, unw_word_t *addr,
