--- libunwind-1.0/src/Makefile.am.orig	2011-08-28 06:58:20.000000000 +0200
+++ libunwind-1.0/src/Makefile.am	2011-09-11 07:52:45.723337644 +0200
@@ -332,7 +332,7 @@
 #
 install-exec-hook:
 	$(LN_S) -f libunwind-$(arch).a  $(DESTDIR)$(libdir)/libunwind-generic.a
-	if test -f $(DESTDIR)$(libdir)/libunwind-generic.so; then \
+	if test -f $(DESTDIR)$(libdir)/libunwind-$(arch).so; then \
 	    $(LN_S) -f libunwind-$(arch).so \
 		$(DESTDIR)$(libdir)/libunwind-generic.so; \
 	fi
